src/twisted/mail/imap4.py

```diff
@@ -209,33 +209,70 @@ def clean(self):
         Clean ranges list, combining adjacent ranges
         """
 
-        self.ranges.sort()
+        ranges = sorted((self._infinity if low is None else low,
+                         self._infinity if high is None else high)
+                        for low, high in self.ranges)
 
-        oldl, oldh = None, None
-        for i,(l, h) in enumerate(self.ranges):
-            if l is None:
+        mergedRanges = [(float('-inf'), float('-inf'))]
+
+
+        for low, high in ranges:
+            previousLow, previousHigh = mergedRanges[-1]
+
+            if previousHigh < low - 1:
+                mergedRanges.append((low, high))
                 continue
-            # l is >= oldl and h is >= oldh due to sort()
-            if oldl is not None and l <= oldh + 1:
-                l = oldl
-                h = max(oldh, h)
-                self.ranges[i - 1] = None
-                self.ranges[i] = (l, h)
 
-            oldl, oldh = l, h
+            mergedRanges[-1] = (min(previousLow, low),
+                                max(previousHigh, high))
 
-        self.ranges = [r for r in self.ranges if r]
+        self.ranges = [(None if low is self._infinity else low,
```

Comment:
I can't help but notice that this fairly complex comprehension occurs twice.  Put it into a function?